src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -1413,9 +1427,17 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
                     // but without anything useful on the call stack. Async Task methods on the other hand return exceptions as the result of the Task.
                     // So it is undesirable to consider these exceptions "user unhandled" since there may well be user code that is awaiting the task.
                     // This is a heuristic since it's possible that there is no user code awaiting the task.
+
+                    // We do the same for async Main methods, since it is unlikely that user code will be awaiting the Task:
+                    // AsyncForwardEntryPoint <Main> -> kick-off method Main -> MoveNext.
+
+                    bool isAsyncMainMoveNext = kickoffMethod.IsEntryPointCandidate &&
+                        entryPointOpt is SynthesizedEntryPointSymbol.AsyncForwardEntryPoint asyncForwardEntryPoint &&
```

Comment:
>SynthesizedEntryPointSymbol.AsyncForwardEntryPoint [](start = 41, length = 50)

Would it make sense to store an instance of this type (or null) on MethodCompiler and pass it to this method instead of a regular MethodSymbol? Then the condition can be simplified to just ```asyncForwardEntryPoint?.UserMain.Equals(kickoffMethod) == true```.

Reply:
Tried this, but it didn't end up simplifying. 
Ended up with multiple method signatures with `SynthesizedEntryPointSymbol.AsyncForwardEntryPoint` and an `as SynthesizedEntryPointSymbol.AsyncForwardEntryPoint` at the start of the chain.

---
In reply to: [280537264](https://github.com/dotnet/roslyn/pull/34548#discussion_r280537264) [](ancestors = 280537264)